Platform/OpenNtfsDxe/Disc.c

```diff
@@ -0,0 +1,731 @@
+  UINTN              Size;
+  EFI_NTFS_FILE      *RootFile;
+
+  Status = DiskRead (FileSystem, 0, sizeof (BOOT_FILE_DATA), &Boot);
+  if (EFI_ERROR(Status)) {
+    return Status;
+  }
+
+  if ((CompareMem (Boot.SystemId, "NTFS", 4) != 0)
+  || (Boot.SectorsPerCluster == 0)
+  || ((Boot.SectorsPerCluster & (Boot.SectorsPerCluster - 1)) != 0)
+  || (Boot.BytesPerSector == 0)
+  || ((Boot.BytesPerSector & (Boot.BytesPerSector - 1)) != 0)) {
+    DEBUG((DEBUG_INFO, "NTFS: (NtfsMount #1) BIOS Parameter Block is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  mSectorSize  = (UINTN) Boot.BytesPerSector;
+  mClusterSize = (UINTN) Boot.SectorsPerCluster * mSectorSize;
+
+  if (Boot.MftRecordClusters > 0) {
+    Size = (UINTN) Boot.MftRecordClusters * mClusterSize;
+  } else if (-Boot.MftRecordClusters >= 31) {
+    DEBUG((DEBUG_INFO, "NTFS: (NtfsMount #2) BIOS Parameter Block is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    Size = (UINTN) LShiftU64 (1ULL, -Boot.MftRecordClusters);
+  }
+
+  mFileRecordSize = Size;
+  if (mFileRecordSize < mSectorSize) {
+    DEBUG((DEBUG_INFO, "NTFS: File Record is smaller than Sector.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  if (Boot.IndexRecordClusters > 0) {
+    Size = (UINTN) Boot.IndexRecordClusters * mClusterSize;
+  } else if (-Boot.IndexRecordClusters >= 31) {
+    DEBUG((DEBUG_INFO, "NTFS: (NtfsMount #3) BIOS Parameter Block is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    Size = (UINTN) LShiftU64 (1ULL, -Boot.IndexRecordClusters);
+  }
+
+  mIndexRecordSize = Size;
+  if (mIndexRecordSize < mSectorSize) {
+    DEBUG((DEBUG_INFO, "NTFS: Index Record is smaller than Sector.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  FileSystem->FirstMftRecord = Boot.MftLcn * mClusterSize;
+
+  /* Driver limitations */
+  if ((mFileRecordSize > NTFS_MAX_MFT) || (mIndexRecordSize > NTFS_MAX_IDX)) {
+    DEBUG((DEBUG_INFO, "NTFS: (NtfsMount #4) BIOS Parameter Block is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  RootFile = AllocateZeroPool (sizeof (EFI_NTFS_FILE));
+  if (RootFile == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  CopyMem (RootFile, &FileSystem->EfiFile, sizeof (EFI_FILE_PROTOCOL));
+
+  RootFile->IsDir         = TRUE;
+  RootFile->Path          = L"/";
+  RootFile->RefCount      = 1;
+  RootFile->FileSystem    = FileSystem;
+  RootFile->RootFile.File = RootFile;
+  RootFile->MftFile.File  = RootFile;
+
+  RootFile->MftFile.FileRecord = AllocateZeroPool (mFileRecordSize);
+  if (RootFile->MftFile.FileRecord == NULL) {
+    FreePool (RootFile);
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Status = DiskRead (
+    FileSystem,
+    FileSystem->FirstMftRecord,
+    mFileRecordSize,
+    RootFile->MftFile.FileRecord
+    );
+  if (EFI_ERROR(Status)) {
+    FreePool (RootFile->MftFile.FileRecord);
+    FreePool (RootFile);
+    return Status;
+  }
+
+  Status = Fixup (RootFile->MftFile.FileRecord, mFileRecordSize, "FILE");
```

Comment:
Mentioned once above. Please rename Fixup to a more meaningful one.

Reply:
https://github.com/acidanthera/OpenCorePkg/pull/332#discussion_r844967230